src/AutoRest.CSharp/Common/Input/Source/ProtocolCompilationInput.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using Microsoft.CodeAnalysis;
+
+namespace AutoRest.CSharp.Input.Source
+{
+    public class ProtocolCompilationInput: CompilationInput
+    {
+        private readonly List<IMethodSymbol> _methodSet = new List<IMethodSymbol>();
+
+        public ProtocolCompilationInput(Compilation compilation)
+            : base(compilation) { }//, type => type.Name.EndsWith("Client"), method => method.Parameters.Length > 0 && method.Parameters.Last().Name == "context") { }
+
+        public override void FilterSymbols()
+        {
+            foreach (IModuleSymbol module in _compilation.Assembly.Modules)
+            {
+                foreach (var type in SourceInputHelper.GetSymbols(module.GlobalNamespace))
+                {
+                    if (type is INamedTypeSymbol typeSymbol && isClient(typeSymbol))
+                    {
+                        foreach (var member in typeSymbol.GetMembers())
+                        {
+                            if (member is IMethodSymbol methodSymbol && isProtocolMethod(methodSymbol))
+                            {
+                                _methodSet.Add(methodSymbol);
+                            }
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
This method is building a cache. We should not bring this into the abstraction. True they all should have this step, but this is not a behavior a `CompilationInput` should have, external members actually should not invoke this method, because this method changes the internal status of this instance.
This cache should be maintained by this instance itself.
Please also refer the below comment as well.

Reply:
Why do you say we change `CompilationInput` to interface?